src/pubsub/src/publisher/publisher.rs

```diff
@@ -118,17 +103,167 @@ impl PublisherBuilder {
     }
 
     pub fn build(self) -> Publisher {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+        // Create the batching worker that will run in the background.
+        // We don't need to keep track of a handle to the worker.
+        // Dropping the Publisher will drop the only sender to the channel.
+        // This will cause worker.run() to read None from the channel and close.
+        let worker = Worker::new(self.topic, self.inner, self.batching_options.clone(), rx);
+        tokio::spawn(worker.run());
+
         Publisher {
-            inner: self.inner,
-            topic: self.topic,
             batching_options: self.batching_options,
+            tx,
+        }
+    }
+}
+
+/// Object that is passed to the worker task over the
+/// main channel. This represents a single message and the sender
+/// half of the channel to resolve the [PublishHandle].
+struct BundledMessage {
+    pub msg: crate::model::PubsubMessage,
+    pub tx: oneshot::Sender<crate::Result<String>>,
+}
+
+/// The worker is spawned in a background task and handles
+/// batching and publishing all messages that are sent to the publisher.
+#[derive(Debug)]
+struct Worker {
+    topic_name: String,
+    client: GapicPublisher,
+    #[allow(dead_code)]
+    batching_options: BatchingOptions,
+    rx: mpsc::UnboundedReceiver<BundledMessage>,
+}
+
+impl Worker {
+    fn new(
+        topic_name: String,
+        client: GapicPublisher,
+        batching_options: BatchingOptions,
+        rx: mpsc::UnboundedReceiver<BundledMessage>,
+    ) -> Self {
+        Self {
+            topic_name,
+            client,
+            rx,
+            batching_options,
+        }
+    }
+
+    async fn run(mut self) {
+        while let Some(msg) = self.rx.recv().await {
+            let client = self.client.clone();
+            let topic = self.topic_name.clone();
+            // In the future, we may also want to keep track of JoinHandles in order to
+            // flush the results.
+            let _handle = tokio::spawn(async move {
+                // For now, we just send the message immediately.
+                // We will want to batch these requests.
+                let request = client
+                    .publish()
+                    .set_topic(topic)
+                    .set_messages(vec![msg.msg]);
+
+                // Handle the response by extracting the message ID on success.
+                let result = request
+                    .send()
+                    .await
+                    .map(|response| response.message_ids.first().cloned().unwrap_or_default());
```

Comment:
Since you are expecting at most 1 element:


```suggestion
                    .map(|response| response.message_ids.pop().unwrap_or_default());
```